src/pymor/algorithms/projection.py

```diff
@@ -63,13 +63,19 @@ def project(op, range_basis, source_basis, product=None):
     assert range_basis is None or range_basis in op.range
     assert product is None or product.source == product.range == op.range
 
-    return ProjectRules(range_basis, source_basis, product).apply(op)
+    rb = product.apply(range_basis) if product is not None and range_basis is not None else range_basis
+
+    try:
+        return ProjectRules(rb, source_basis).apply(op)
+    except NoMatchingRuleError:
+        op.logger.warning('Using inefficient generic projection operator')
+        return ProjectedOperator(op, range_basis, source_basis, product)
```

Comment:
I guess this would make a difference if somebody used `ProjectRules` directly. Is this something that could happen?

Reply:
In principle you might want to do that to benefit from caching of projected `Operators` when the same `Operator` appears multiple times in the `Model` you are projecting. However, we currently don't do that in pyMOR and the `ProjectedOperator` case is something that usually should not apply. So I would say it is ok to remove it from `ProjectRules`.